043-tasks_queue/proposal.md

```diff
@@ -0,0 +1,145 @@
+
+As Concourse operators, today we see that Concourse's scheduling process is:
+
+1. not safe: workers are not protected from overload. When the system is overloaded it has no chance to recover.
+2. not fair: tasks are racing for being executed without any specific order
+3. not scalable: it is not possible to measure if the number of workers is appropriate to the current user demand.
+
+To highlight the problem in more detail here's an overview of Concourse's task scheduling process.
+
+1. the module scheduler selects the tasks that can be run (all their inputs are available, their job/pipeline is not paused)
+2. the task submits itself to the WorkerPool
+3. the WorkerPool selects a compatible worker according to the configured PlacementStrategy
+4. the task runs
+
+During this process no component asks whether the system, in particular the workers, are capable of accepting more tasks. The different placement strategy implementations only express preferences.
+
+The scheduling process requires a new component which keeps track of the available resources as tasks start, execute and stop on the available workers.
+
+
+# Proposal
+
+Let's take a look at the RunTaskStep function from (`atc/worker/client.go`)
+
+```go
+func (client *client) RunTaskStep(...) TaskResult {
+    //...
+    chosenWorker, err := client.chooseTaskWorker(...)
+    //...
+}
+```
+This function is called from the `Run()` method of TaskStep, which is effectively responsible for starting the task's container.
+
+The task chooses a worker given the placement strategy and the worker starts executing the corresponding container. This happens without taking into account the available resources on the worker. If users submit more jobs than the workers can instantaneously handle, the system becomes overloaded. Additionally, in case of overload no corrective action is possible since the tasks cannot re-schedule themselves.
+
+Instead of directly allocating a worker from a task, we propose the following, extended worker allocation process:
+
+![resource pool schema](https://ci-pix4d-internal-static.s3-eu-west-1.amazonaws.com/concourse-scheduling-rfc.png)
+
+A new component, called Resource Pool is placed between the worker client (the ATC) and worker pool. The client would interact with the worker pool as follows:
+
+1. The client requests a worker from Resource Pool.  This call blocks until a worker becomes free.
+2. The resource pool returns the worker to the client.  The worker has been chosen from the available workers using the container placement strategy.
+3. The client starts the task container on the allocated worker
+4. The task container stops
+5. The worker client releases the worker's resources
```

Comment:
I would rather that Workers "pull" from tasks that are ready to run, rather than the task itself blocking on finding a worker.

I guess this is why the name is "Resource Pool", not "Tasks Pool".

Reply:
This would be a lot more complex than it sound: currently the workers (in my understanding at least), are quite blind about the tasks and resources that they are required to process.
So first of all we would need to make the workers aware of the workload they have to process and then let them manage it.
That solution would also make it a lot harder to have meaningful metrics about the tasks (those waiting and those running).
In the end I don't think there is really nothing to gain from the "pull" architecture as the ATC will always be there as the central dispatch and point of contact of the DB where, in any case, the state of all tasks must be recorded.